File: ioi2020-biscuits/solution/biscuits-yanhao-smallX_top_down.cpp
#include "biscuits.h"
#include <vector>
#include <unordered_map>
using namespace std;

const int x_max = 3e5+1;
long long dp[61][2*x_max];

long long f(int idx, long long w, long long x, vector<long long> &a) {
    if(idx==60) return 1;
    if(dp[idx][w]!=-1) return dp[idx][w];
    return dp[idx][w] = f(idx+1, (w+a[idx])/2, x, a) + (w+a[idx]>=x ? f(idx+1, (w+a[idx]-x)/2, x, a) : 0);
}

void merge_biscuits(long long x, vector<long long> &a) {
    while(a.size()<61) a.push_back(0);
    for(int i=0; i<(int)a.size(); i++) {
        if(a[i]>x+1) {
            a[i+1] += (a[i]-x)/2;
            a[i] -= ((a[i]-x)/2)*2;
        }
    }
}

long long count_tastiness(long long x, std::vector<long long> a) {
    for(int i=0; i<=60; i++) {
        for(int j=0; j<=2*x; j++) {
            dp[i][j] = -1;
        }
    }
    merge_biscuits(x, a);
    return f(0, 0, x, a);
}

File: ioi2020-biscuits/solution/biscuits-yanhao-smallX_top_down.java
class biscuits {

	final int x_max = (int)1e5+1;
	long[][] dp = new long[61][2*x_max];

	long f(int idx, long w, long x, long[] a) {
		if(idx==60) return 1;
		if(dp[idx][(int)w]!=-1) return dp[idx][(int)w];
		return dp[idx][(int)w] = f(idx+1, (w+a[idx])/2, x, a) + (w+a[idx]>=x ? f(idx+1, (w+a[idx]-x)/2, x, a) : 0);
	}
	
	long[] merge_biscuits(long x, long[] a) {
		long[] b = new long[61];
		for(int i=0; i<=60; i++) {
			b[i] = (i < a.length ? a[i] : 0);
		}
		for(int i=0; i<b.length; i++) {
			if(b[i]>x+1) {
				b[i+1] += (b[i]-x)/2;
				b[i] -= (b[i]-x)&(-2);
			}
		}
		return b;
	}

	long count_tastiness(long x, long[] a) {
		a = merge_biscuits(x, a);
		for(int i=0; i<=60; i++) {
			for(int j=0; j<=2*x; j++) {
				dp[i][j] = -1;
			}
		}
		return f(0, 0, x, a);
	}
}